### ISSUE TITLE
[Idea] Alternative API for Blockwise io_deps

### ISSUE THREAD

**Issue by gjoseph92** [dask/dask#7513]
tl;dr: what if we extend `Blockwise` to also take array-like objects in its inputs, along with symbols? Think of it as "move `io_deps` into the `indices`", with a few advantages that fall out:
1) More consistent interface to Blockwise: think about IO in terms of broadcasting (auxiliary) data against a chunk pattern, just like Blockwise already does.
1) More explicit pattern for serialization and managing opaque serialized data on the scheduler. Also, we can cull unneeded parts of the array-like data prior to serialization.
1) Separation of concerns: `io_deps` currently mixes serialization info (module path) with the blockwise arguments, and `make_blockwise_graph` includes deserialization of the `BlockwiseIODeps`. This fully separates graph logic from serialization.

Proposal:
What if, instead of just being able to give symbolic arguments (like `"x"`, `"ij"`) to Blockwise, you could also give array-like things directly (like `np.arange(10)`, `"i"`). By “array-like” I mean a concept similar to [`BlockwiseIODeps`](https://github.com/dask/dask/blob/c5633c21e754df6b2c13195f1e2688e347c48963/dask/blockwise.py#L33-L48): any object with a `shape` that supports multidimensional indexing. Could be an ndarray, or could be a class with a `__getitem__`. It would let you do stuff like:
```python
    >>> y = np.arange(3)
    >>> make_blockwise_graph(add, 'z', 'i', 'x', 'i', y, 'i', numblocks={"x": (3,)})
    ('z', 0): (add, ('x', 0), 0)
    ('z', 1): (add, ('x', 1), 1)
    ('z', 2): (add, ('x', 2), 2)

    >>> y = np.arange(6).reshape(3, 2)
    >>> make_blockwise_graph(np.outer, 'z', 'ij', 'x', 'i', y, 'j', numblocks={"x": (2,)})
    ('z', 0, 0): (<function outer at 0x7fa6c8923e50>, ('x', 0), array([0, 1]))
    ('z', 0, 1): (<function outer at 0x7fa6c8923e50>, ('x', 0), array([2, 3]))
    ('z', 0, 2): (<function outer at 0x7fa6c8923e50>, ('x', 0), array([4, 5]))
    ('z', 1, 0): (<function outer at 0x7fa6c8923e50>, ('x', 1), array([0, 1]))
    ('z', 1, 1): (<function outer at 0x7fa6c8923e50>, ('x', 1), array([2, 3]))
    ('z', 1, 2): (<function outer at 0x7fa6c8923e50>, ('x', 1), array([4, 5]))
```
(notice how `y` is getting indexed at graph-construction time, and the contents are inlined into tasks. Just like we do with `io_deps`, you could do fancier things with that interface than just inline data: have `__getitem__` return a block info dict, or generate a sub-task for the current block coordinates, etc.)

To support serialization, these array-like objects would have to have a `__dask_distributed_pack__` and `__dask_distributed_unpack__` method. For convenience, any plain object (like a NumPy array) would be wrapped in [`BlockwiseNdarray`](https://github.com/dask/dask/commit/a154dae7f9893965a95ecd3dde6ec0fa6868764c#diff-b2b064ba4d14c2c4d3c14bb57d0cda3849f084625015bb61f2ec1f7ffe8195ccR229-R256) for you [by the `Blockwise` constructor](https://github.com/dask/dask/commit/a154dae7f9893965a95ecd3dde6ec0fa6868764c#diff-b2b064ba4d14c2c4d3c14bb57d0cda3849f084625015bb61f2ec1f7ffe8195ccR378-R380).

The `__dask_distributed_pack__` interface is more specific than the current `BlockwiseIODeps`: it's [passed the list of indices](https://github.com/dask/dask/blob/a154dae7f9893965a95ecd3dde6ec0fa6868764c/dask/blockwise.py#L501-L520) into that argument that will actually be used (after culling). (With current `io_deps`, I don't believe each `BlockwiseIODeps` gets this chance to drop unnecessary data before packing.) Using this list of indices, we can pre-slice out only the necessary chunks (of an ndarray, for example) on the client, serialize them, and stick them in a dict that supports equivalent `__getitem__`s (example [here](https://github.com/dask/dask/commit/a154dae7f9893965a95ecd3dde6ec0fa6868764c#diff-b2b064ba4d14c2c4d3c14bb57d0cda3849f084625015bb61f2ec1f7ffe8195ccR237-R256)). So when generating the graph on the scheduler, we just select those opaque `Serialized` blobs and inline them into the graph. Then on the worker, the `Serialized` objects are unpacked. (This `Serialized`-objects-on-the-scheduler part is the same as our plan with `BlockwiseIODeps` right now; in both cases, this is aspirational, since we need to adjust `dumps_task` serialization to make this happen.)

This all is nearly isomorphic with the current `io_deps`. The reason it appeals to me is that conceptually, `Blockwise` is already all about broadcasting array-ish things together. This just extends that concept to more objects while using the same interface. For me at least, it feels less like understanding a “new thing” than `io_deps`, which could be helpful for other contributors/users.

I also feel like it would shift my mindset when implementing IO operations towards “how do I get this auxiliary data (CSV filenames, `np.random` seeds, etc.) into the right shape so it broadcasts against my output chunk pattern, then let blockwise figure out the mappings?” Whereas to me, `io_deps` encourages thinking about “how to I map from a coordinate in the output to the inputs that task needs,” which conceptually I’ve found harder: normally, it's blockwise’s job to figure out those mappings for you.

I think nice IO patterns emerge given this interface and a few helpers to tell your function where it is within the block structure (like [`Slices`](https://github.com/dask/dask/blob/a154dae7f9893965a95ecd3dde6ec0fa6868764c/dask/blockwise.py#L259-L276) and [`BlockIndex`](https://github.com/dask/dask/blob/a154dae7f9893965a95ecd3dde6ec0fa6868764c/dask/blockwise.py#L279-L288)). Instead of making subclasses of `BlockwiseIODeps` for different IO operations, you might be able to compose together those helpers, a reshaped ndarray of your inputs, and a function for many IO cases.

I partially implemented this at https://github.com/dask/dask/compare/main...gjoseph92:blockwise-arraylike-args?expand=1. For a smaller diff, this starts just prior to #7281. I recommend only looking at the commits for:
* [`make_blockwise_graph` changes](https://github.com/dask/dask/commit/3b10f23672b7221e0c8ac22262319a795869125f)
* [pack/unpack logic](https://github.com/dask/dask/commit/a154dae7f9893965a95ecd3dde6ec0fa6868764c)
* most useful, an (I think complete?) [implementation of `Array.random`](https://github.com/dask/dask/commit/ff15a85311c778636578234f3a37fee1cedba4a0) using this

Note that I haven't implemented blockwise rewriting here; intuitively I think it should work, but it's currently broken.

cc @rjzamora @quasiben @jrbourbeau @ian-r-rose @mrocklin

**Comment by jakirkham** [comment]
cc @madsbk

**Comment by rjzamora** [comment]
Thanks for this @gjoseph92!  As I mentioned offline, I am a mostly supportive of this change.  I don’t think the actual functionality changes at all, and it will require slightly more changes to Blockwise than io_deps required, but future development/maintenance should become more intuitive.

> The __dask_distributed_pack__ interface is more specific than the current BlockwiseIODeps: it's passed the list of indices into that argument that will actually be used (after culling). (With current io_deps, I don't believe each BlockwiseIODeps gets this chance to drop unnecessary data before packing.) Using this list of indices, we can pre-slice out only the necessary chunks (of an ndarray, for example) on the client, serialize them, and stick them in a dict that supports equivalent __getitem__s (example here).

 It is certainly possible to achieve the same behavior with `io_deps` if we, for example,  use `output_blocks` to cull the `io_deps` dictionary (or `BlockwiseIODeps` object) in the `Blockwise.__init__`.  I needed to deal with something similar when I was recently experimenting with DataFrame column-selection, so I am aware that the ugliness here is an io_deps weakness.    Altough it **can** be impelemented without the changes you are proposing, I do agree that most of the Blockwise + IO logic becomes more intuitive (and may require less case-specific logic) when the IO-argument object is wrapped in a class that guarentees desireable indexing behavior, and when the underlying indexable object is automatically wrapped within `Blockwise` itself.  With that change in place, there is no functional difference between the `indices` and `io_deps` approaches.   Therefore, if the IO dependencies **can** be included in `indices`, I agree that the general API becomes cleaner :)
 
 Since `Blockwise` is generally intended for operating on array-like things, I think it is okay to use array-specific language.  However, I recommend that we design all basic-IO functionality to work with a class specification that is not **too** array specific.  That is, it would be nice to allow the “indexable”-element that is passed in with `indices` to be other objects that can be indexed with `getitem` (e.g. array, dataframe/series, and dict). So, rather than integrating directly with `ArrayLike`/`BlockwiseArraylike`/`BlockwiseNdarray`, we probably need the base `Blockwise` logic to work with something like `Indexable`/`BlockwiseIndexable`.

**Comment by gjoseph92** [comment]
> It is certainly possible to achieve the same behavior with `io_deps`

Agreed. To be clear, besides the less-tangible benefits of the interface change, the other benefits I listed could be backported to `io_deps`. As you noted, requiring IO arguments to follow a specific interface, where both `__getitem__` and `__dask_distributed_*pack__` are well-defined, is really the important change here. Once you've done that, the distinction between `blockwise(..., "x", "ij", io_deps={"x"; x_obj})` and `blockwise(..., x_obj, "ij")` is more spelling and style.

>  So, rather than integrating directly with `ArrayLike`/`BlockwiseArraylike`/`BlockwiseNdarray`, we probably need the base `Blockwise` logic to work with something like `Indexable`/`BlockwiseIndexable`.

Yeah, `Indexable` (or `Mapping` or something) is probably a better term here—`ArrayLike` is a bit misleading, and sounds more specific than it actually is. It doesn't actually have to be an array; anything with `__getitem__` works. But the important restriction beyond any old `Mapping` is that the keys of your `Indexable` need to be tuples of ints of length N, where N is however many dimensions you told Blockwise the object has via the indices (like `"ijk"` means 3D). Also, the object needs to have a `shape` attribute (could be renamed), just because there's no way to refer to that object in the `numblocks` dict. So between those two restrictions, the _interface_ feels a little ndarray-like, but the implementation doesn't have to involve arrays at all.

In fact, if we didn't like the requirement that `Indexable`/`ArrayLike` arguments had to have a `shape` attribute, that's a reason to keep `io_deps`: you'd still give a symbolic name for your array-like, then refer to it by that in both `numblocks` and `io_deps`.

**Comment by rjzamora** [comment]
>But the important restriction beyond any old Mapping is that the keys of your Indexable need to be tuples of ints of length N, where N is however many dimensions you told Blockwise the object has via the indices (like "ijk" means 3D). Also, the object needs to have a shape attribute (could be renamed), just because there's no way to refer to that object in the numblocks dict. So between those two restrictions, the interface feels a little ndarray-like, but the implementation doesn't have to involve arrays at all.

Yeah - On second thought, it is probably fine to focus on array input for automatic `BlockwiseNdarray` creation within `Blockwise`.  It makes sense to require other `BlockwiseArraylike` objects to be defined outside `Blockwise.__init__` when the logic is bit trickier (because `shape` and `__getitem__` may not be "pass-through" operations).

**Comment by ian-r-rose** [comment]
I like this idea very much.

>     1. More consistent interface to Blockwise: think about IO in terms of broadcasting (auxiliary) data against a chunk pattern, just like Blockwise already does.

This, in particular, would help with some of the issues that I have been thinking about in #7417 where some code paths for array creation might involve a (lazy) array, and others might not, determining whether we have to go down a `Blockwise` or `BlockwiseIO` code path. Having a bunch of `if` branches handling all of those is much more difficult to reason about than making sure the input indices match a protocol. As @rjzamora says, I don't think it changes what is possible, but is easier to reason about.

I think we'd have to think very carefully about the specific API for the `ArrayLike` protocol. I like the idea of using `Protocol`s and `Generic`s for it (though I'd note that `mimic_numblocks` uses `arr.reshape`, which is not available). It got me thinking of whether it would be possible to define a msgpack-serializable bound on `StateT` (though maybe not possible with current python typing, since it has struggled with recursive types).

**Comment by gjoseph92** [comment]
@ian-r-rose did you look at https://github.com/dask/dask/commit/ff15a85311c778636578234f3a37fee1cedba4a0? Since I cribbed a lot of ideas from #7417, I'm particularly curious what you think. Being able to avoid the `if`s and handle both lazy and non-lazy arrays was quite nice.

> I'd note that mimic_numblocks uses arr.reshape, which is not available

Yeah, this should have a more specific tighter type bound. I would have just required `np.ndarray`, but didn't want to add the import. On that note, I doubt that the file organization on my branch is what we'd actually want to have in the end—for example, `mimic_numblocks` might belong in `array/blockwise.py`?

> It got me thinking of whether it would be possible to define a msgpack-serializable bound on `StateT`

I almost did that! But I thought it'd be overkill for this draft PR. Pretty sure it's possible; I've done it with JSON before which is basically the same.

> I like the idea of using `Protocol`s and `Generic`s for it

Small note: `Protocol` was introduced in 3.8, so we'd need to add https://pypi.org/project/typing-extensions/ as a dependency for 3.7 in order to use it. Curious if that would be okay, or if typing isn't enough of a priority in dask to justify it?

**Comment by ian-r-rose** [comment]
> @ian-r-rose did you look at [ff15a85](https://github.com/dask/dask/commit/ff15a85311c778636578234f3a37fee1cedba4a0)? Since I cribbed a lot of ideas from #7417, I'm particularly curious what you think. Being able to avoid the `if`s and handle both lazy and non-lazy arrays was quite nice.

Yes, I did, and I like it very much :+1:  . I think it's a nice improvement.

> 
> > I'd note that mimic_numblocks uses arr.reshape, which is not available
> 
> Yeah, this should have a more specific tighter type bound. I would have just required `np.ndarray`, but didn't want to add the import. On that note, I doubt that the file organization on my branch is what we'd actually want to have in the end—for example, `mimic_numblocks` might belong in `array/blockwise.py`?

Sure, makes sense. This is essentially the same issue that we need to finesse in other places: how do we support ndarrays without importing numpy?  I would be to go ahead and put the utility function in `dask.array.utils` or similar where it can reference numpy, since that's the intent anyways.

> 
> I almost did that! But I thought it'd be overkill for this draft PR. Pretty sure it's possible; I've done it with JSON before which is basically the same.

I know it's doable with Typescript, which has better recursive type definitions. I don't think it's easy (and maybe not possible at all) in Python, cf https://github.com/python/typing/issues/182

> > I like the idea of using `Protocol`s and `Generic`s for it
> 
> Small note: `Protocol` was introduced in 3.8, so we'd need to add https://pypi.org/project/typing-extensions/ as a dependency for 3.7 in order to use it. Curious if that would be okay, or if typing isn't enough of a priority in dask to justify it?

I know that @crusaderky has advocated for that. I'd be interested in having access to it as well, though opinions differ

**Comment by crusaderky** [comment]
> so we'd need to add https://pypi.org/project/typing-extensions/ as a dependency for 3.7 in order to use it

You actually need it for all versions, or otherwise your imports will be a forest of ```try... except ImportError``. Also mypy is not always brilliant on conditional imports (remember, it does not run the code).

**Comment by ian-r-rose** [comment]
This is largely done. There are some things which could probably make it a bit nicer to work with, but I'm not sure it's worth the time to put effort into `Blockwise` right now. Do you think we can safely close this @gjoseph92 and @rjzamora?

**Comment by gjoseph92** [comment]
> requiring IO arguments to follow a specific interface, where both `__getitem__` and `__dask_distributed_*pack__` are well-defined, is really the important change here
https://github.com/dask/dask/issues/7513#issuecomment-813596159

I this this all has now been done. 

> Once you've done that, the distinction between `blockwise(..., "x", "ij", io_deps={"x"; x_obj})` and `blockwise(..., x_obj, "ij")` is more spelling and style.

And this spelling is mostly supported now too.

So yeah, I'd say we're basically here for the core functionality! It can still feel pretty cumbersome to use, but I think I agree that we're not likely to work on it anytime soon, so we might as well close this?

**Comment by rjzamora** [comment]
>So yeah, I'd say we're basically here for the core functionality! It can still feel pretty cumbersome to use, but I think I agree that we're not likely to work on it anytime soon, so we might as well close this?

Yeah, I think it makes sense to close this. There is certainly room for improvement, but the basic functionality is there.

Others should feel free to re-open if I am missing something.
